### PR TITLE
BL-3307 Make immediately needed thumbnails in foreground

### PR DESCRIPTION
**Pull request by JohnThomson**
This change is [Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/1054)

### CODE REVIEW

**Comment by hatton** [comment]
Reviewed 5 of 6 files at r1.
Review status: 5 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/BloomExe/BookThumbNailer.cs, line 161 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1054#-KHN47sxXPKtIWF3bkih:-KHN47sxXPKtIWF3bkii:1773100723) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/4322129457ad2c88d1149186cc55d2536b8a8295/src/BloomExe/BookThumbNailer.cs#L161)):*
> ```Smalltalk
> 		/// <param name="callback"></param>
> 		/// <param name="errorCallback"></param>
> 		public void RebuildThumbNailAsync(Book.Book book, HtmlThumbNailer.ThumbnailOptions thumbnailOptions,
> ```

I don't know if it's worth re-doing, but I wonder if things wouldn't be simpler by making what you have here the "inner" method, and then have two public methods. One with the same name and signature as before, and a new one, RebuildThumbnailNow(), that is non async. That way calling code would be easier to read and reason about. I find bools to be kind of hidden information.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1054)*

**Comment by JohnThomson** [comment]
Review status: 5 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/BloomExe/BookThumbNailer.cs, line 161 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1054#-KHN47sxXPKtIWF3bkih:-KHPv3suZ4mSpYErwQJS:1795614352) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/4322129457ad2c88d1149186cc55d2536b8a8295/src/BloomExe/BookThumbNailer.cs#L161)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> I don't know if it's worth re-doing, but I wonder if things wouldn't be simpler by making what you have here the "inner" method, and then have two public methods. One with the same name and signature as before, and a new one, RebuildThumbnailNow(), that is non async. That way calling code would be easier to read and reason about. I find bools to be kind of hidden information.</details>
I went back and forth on this one. The problem is, there are three or four layers of method, each of which would have two mostly-identical implementations except that one calls the async and one the non-async method at the next layer down.

Maybe a compromise would be to keep the current method public; then add the Async and Now versions for clearer use by code that knows which it wants.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1054)*

**Comment by JohnThomson** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/BloomExe/BookThumbNailer.cs, line 161 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1054#-KHN47sxXPKtIWF3bkih:-KHR4VqUpDx4HarUFHx4:-936584293) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/4322129457ad2c88d1149186cc55d2536b8a8295/src/BloomExe/BookThumbNailer.cs#L161)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> I went back and forth on this one. The problem is, there are three or four layers of method, each of which would have two mostly-identical implementations except that one calls the async and one the non-async method at the next layer down.
> 
> Maybe a compromise would be to keep the current method public; then add the Async and Now versions for clearer use by code that knows which it wants.</details>
OK, I think I've done what you want. I've made Async and Now versions wherever a caller wanted to pass an explicit true or false, and made the neutral version private where possible.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1054)*

File: src/BloomExe/BookThumbNailer.cs [hunk 1]
```diff
@@ -152,7 +151,34 @@ public Image GetThumbnailForPage(Book.Book book, IPage page, bool isLandscape)
...
+		{
+			RebuildThumbNail(book, thumbnailOptions, callback, errorCallback, true);
+		}
+
+		/// <summary>
+		/// Will call either 'callback' or 'errorCallback' UNLESS the thumbnail is readonly, in which case it will do neither.
+		/// </summary>
+		/// <param name="book"></param>
+		/// <param name="thumbnailOptions"></param>
+		/// <param name="callback"></param>
```
Comment (hatton) [inline src/BloomExe/BookThumbNailer.cs]: Can you add a comment saying what the callback is for? I wouldn't expect this on a synchronous method.

**Comment by JohnThomson** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/BloomExe/BookThumbNailer.cs, line 165 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1054#-KH6nhZERahhwxJyiwQc-r2-165:-KHRGO9cFUg5Uv5_MtaN:589841448) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a7feb5d48acfa888af394789d7d7423cb5b3f95d/src/BloomExe/BookThumbNailer.cs#L165)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> Can you add a comment saying what the callback is for? I wouldn't expect this on a synchronous method.</details>
I removed it. Single client has no need of it. Was only there for consistency across the similar methods.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1054)*

**Comment by hatton** [comment]
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/BloomExe/BookThumbNailer.cs, line 161 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1054#-KHRNKtKAeKJYMylZ9On:-KHRNKtKAeKJYMylZ9Oo:1903459714) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/13a5caa5329ac106145ab4718ba8a18dc20b87e4/src/BloomExe/BookThumbNailer.cs#L161)):*
> ```Smalltalk
> 			Action<BookInfo, Exception> errorCallback)
> 		{
> 			RebuildThumbNail(book, thumbnailOptions, (info, image) => { }, errorCallback, false);
> ```

Same with the other callback. Synchronous methods can throw, or return, but not do callbacks. Since your only client here supplies a callback that just throws, migh as well remove this parameter and supply your own callback that does the throwing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1054)*

**Comment by JohnThomson** [comment]
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/BloomExe/BookThumbNailer.cs, line 161 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1054#-KHRNKtKAeKJYMylZ9On:-KHeeXDCGbWIrwO-dV2t:-499146009) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/13a5caa5329ac106145ab4718ba8a18dc20b87e4/src/BloomExe/BookThumbNailer.cs#L161)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> Same with the other callback. Synchronous methods can throw, or return, but not do callbacks. Since your only client here supplies a callback that just throws, migh as well remove this parameter and supply your own callback that does the throwing.</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1054)*

**Pull request merged**